File: python/agent/visionread.py
# -*- coding: utf-8 -*-
"""用**视觉模型**读扫描页 —— 把 PDF 每一页渲染成图，直接让模型读。

为什么有这条路：
    扫描版 PDF 没有文字层，本地解析器一个字都抽不出来。传统解法是本地 OCR
    （MinerU 之类，要下几百 MB~几 GB 模型）。但如果手上的模型本身就能看图，
    **把页面当图片发过去**就行了 —— 不用下 OCR 模型，表格、图表、手写也能读。

两条路各有取舍：
    · 本地 OCR（MinerU）  全本地、一次装好；质量取决于模型
    · 视觉模型读页        零本地模型、立刻可用；按页调用（本地 Ollama 免费，
                          云端按 token 计费），**页面会发到模型服务商**

所以这里做成**可选**：配了才走，没配就完全不参与。

配置（放在本机的 .env 里 —— 这是**每台机器自己**的事，不要提交）：
    GALGAME_VISION_BASE    OpenAI 兼容地址
                           本地 Ollama：http://127.0.0.1:11434/v1
                           DeepSeek   ：https://api.deepseek.com
    GALGAME_VISION_MODEL   模型名，如 qwen3.5:4b / deepseek-chat
    GALGAME_VISION_KEY     API Key（本地 Ollama 不需要，留空即可）
    GALGAME_VISION_DPI     渲染精度，默认 150（越高越清楚也越慢）
    GALGAME_VISION_MAX_PAGES  单本最多读多少页，默认 40
    GALGAME_VISION_PROMPT  自定义提示词（可选）

渲染需要 pypdfium2（一个 ~3MB 的小轮子）。**没装的话这一路会自动跳过**，
行为与以前一模一样。
"""

import base64
import json
import os
import shutil
import subprocess
import sys
import tempfile
import urllib.error
import urllib.request

DEFAULT_DPI = int(os.environ.get('GALGAME_VISION_DPI') or 150)
DEFAULT_MAX_PAGES = int(os.environ.get('GALGAME_VISION_MAX_PAGES') or 40)
PAGE_TIMEOUT = int(os.environ.get('GALGAME_VISION_TIMEOUT') or 180)

# 让模型只吐正文，别加解释 —— 抽取不是聊天
DEFAULT_PROMPT = (
    '把这张书页图片里的文字**完整、逐字**转写成纯文本，保持原有段落顺序。\n'
    '要求：\n'
    '· 只输出正文，不要任何解释、不要加「以下是」「这张图片」之类的话；\n'
    '· 不要翻译、不要改写、不要总结；\n'
    '· 看不清的字用 □ 代替，不要猜；\n'
    '· 如果整页没有文字（比如纯插图页），只回一个空行。'
)


def vision_config():
    """读取视觉模型配置。"""
    return {
        'base': (os.environ.get('GALGAME_VISION_BASE') or '').strip().rstrip('/'),
        'model': (os.environ.get('GALGAME_VISION_MODEL') or '').strip(),
        'key': (os.environ.get('GALGAME_VISION_KEY') or '').strip(),
        'dpi': DEFAULT_DPI,
        'maxPages': DEFAULT_MAX_PAGES,
        'prompt': os.environ.get('GALGAME_VISION_PROMPT') or DEFAULT_PROMPT,
    }


def vision_configured():
    c = vision_config()
    return bool(c['base'] and c['model'])


def renderer_available():
    """页面渲染器在不在（pypdfium2）。"""
    try:
        import pypdfium2  # noqa: F401
        return True
    except Exception:
        return False


def vision_available():
    """这一路能不能用：配置齐 + 渲染器在。"""
    return vision_configured() and renderer_available()


def status():
    """给自检/诊断用。"""
    c = vision_config()
    return {
        'configured': vision_configured(),
        'renderer': renderer_available(),
        'ready': vision_available(),
        'base': c['base'] or None,
        'model': c['model'] or None,
        'dpi': c['dpi'],
        'maxPages': c['maxPages'],
    }


# ══════════════════════════════════════════════════════════════════
#  渲染：PDF → 每页一张 PNG
# ══════════════════════════════════════════════════════════════════

def render_pdf(data, dpi=None, max_pages=None):
    """把 PDF 渲染成 PNG 列表。返回 (pages, note)：pages 是 bytes 列表。

    需要 pypdfium2；没装会抛 RuntimeError，由上层如实告诉用户。
    """
    try:
        import pypdfium2 as pdfium
    except Exception:
        raise RuntimeError('缺少页面渲染器 pypdfium2（pip install pypdfium2）')

    dpi = dpi or DEFAULT_DPI
    max_pages = max_pages or DEFAULT_MAX_PAGES
    scale = dpi / 72.0

    out = []
    pdf = pdfium.PdfDocument(data)
    try:
        n = len(pdf)
        for i in range(min(n, max_pages)):
            page = pdf[i]
            try:
                bitmap = page.render(scale=scale)
                pil = bitmap.to_pil()
                buf = _png_bytes(pil)
                out.append(buf)
            finally:
                try:
                    page.close()
                except Exception:
                    pass
    finally:
        try:
            pdf.close()
        except Exception:
            pass
    return out, ('共 %d 页，只读了前 %d 页' % (n, max_pages) if n > max_pages else '')


def _png_bytes(pil_image):
    """PIL 图 → PNG bytes（不依赖 Pillow 之外的任何东西）。"""
    import io
    bio = io.BytesIO()
    pil_image.save(bio, format='PNG', optimize=True)
    return bio.getvalue()


# ══════════════════════════════════════════════════════════════════
#  调用视觉模型
# ══════════════════════════════════════════════════════════════════

def transcribe_page(png_bytes, cfg=None, timeout=None):
    """把一页图片交给视觉模型，返回它读出来的文字。"""
    c = cfg or vision_config()
    if not (c['base'] and c['model']):
        raise RuntimeError('视觉模型未配置')

    b64 = base64.b64encode(png_bytes).decode('ascii')
    body = {
        'model': c['model'],
        'messages': [{
            'role': 'user',
            'content': [
                {'type': 'text', 'text': c['prompt']},
                {'type': 'image_url', 'image_url': {'url': 'data:image/png;base64,' + b64}},
            ],
        }],
        'temperature': 0,          # 抽取任务要确定性，不要发挥
        'max_tokens': 8000,
    }
    headers = {'content-type': 'application/json'}
    if c['key']:
        headers['authorization'] = 'Bearer ' + c['key']

    req = urllib.request.Request(
        c['base'] + '/chat/completions',
        data=json.dumps(body, ensure_ascii=False).encode('utf-8'),
        headers=headers, method='POST')

    try:
        with urllib.request.urlopen(req, timeout=timeout or PAGE_TIMEOUT) as resp:
            payload = json.loads(resp.read().decode('utf-8'))
    except urllib.error.HTTPError as e:
        detail = ''
        try:
            detail = e.read().decode('utf-8', 'replace')[:300]
        except Exception:
            pass
        raise RuntimeError('视觉模型返回 HTTP %s：%s' % (e.code, detail))
    except Exception as e:
        raise RuntimeError('连不上视觉模型 %s（%s）' % (c['base'], e))

    choices = payload.get('choices') or [{}]
    msg = (choices[0] or {}).get('message') or {}
    return (msg.get('content') or '').strip()


# ══════════════════════════════════════════════════════════════════
#  对外入口
# ══════════════════════════════════════════════════════════════════

def read_pdf(data, filename='', on_progress=None):
    """用视觉模型读整本 PDF。返回 (text, note)。

    note 非空表示不完整或失败 —— 由上层如实告诉用户，而不是假装读到了。
    """
    if not vision_configured():
        return '', '视觉模型未配置'
    if not renderer_available():
        return '', '缺少页面渲染器 pypdfium2（pip install pypdfium2）'

    c = vision_config()
    try:
        pages, page_note = render_pdf(data, c['dpi'], c['maxPages'])
    except Exception as e:
        return '', '渲染 PDF 失败：%s' % e
    if not pages:
        return '', '这本 PDF 渲染不出任何页面'

    parts = []
    failed = 0
    for i, png in enumerate(pages):
        try:
            t = transcribe_page(png, c)
        except Exception as e:
            failed += 1
            if on_progress:
                on_progress(i + 1, len(pages), '失败：%s' % e)
            continue
        if t:
            parts.append(t)
        if on_progress:
            on_progress(i + 1, len(pages), '%d 字' % len(t))

    text = '\n\n'.join(parts)
    if not text.strip():
        return '', '视觉模型没读出任何文字（读了 %d 页）' % len(pages)
    note = page_note
    if failed:
        note = (note + '；' if note else '') + '有 %d 页读取失败' % failed
    return text, note


def selftest():
    """自检用：报告这一路的状态，能连就顺手试一页。"""
    st = status()
    if not st['ready']:
        st['note'] = ('未配置 GALGAME_VISION_BASE/MODEL' if not st['configured']
                      else '缺少 pypdfium2（pip install pypdfium2）')
        return st
    # 造一张只有一行字的图，看看模型能不能读出来
    try:
        png = _probe_image()
        got = transcribe_page(png, timeout=60)
        st['probe'] = got[:60]
        st['ok'] = '测试' in got or 'OK' in got.upper()
    except Exception as e:
        st['ok'] = False
        st['probe_error'] = str(e)
    return st


def _probe_image():
    """生成一张写着「测试」的 PNG —— 用 PIL 画，不依赖系统字体。"""
    try:
        from PIL import Image, ImageDraw
    except Exception:
        raise RuntimeError('自检需要 Pillow')
    img = Image.new('RGB', (240, 90), 'white')
    d = ImageDraw.Draw(img)
    d.rectangle([5, 5, 235, 85], outline='black', width=2)
    d.text((24, 34), 'VISION OK 12345', fill='black')
    import io
    bio = io.BytesIO()
    img.save(bio, format='PNG')
    return bio.getvalue()

File: workbuddy/skills/academic-galgame/scripts/visionread.js
/**
 * 视觉读页 —— 让「有眼睛的模型」直接看扫描书的页面。
 *
 * 为什么需要这一路：
 *   扫描件之所以读不出字，是因为整页就是**一张图**。与其费劲做 OCR，
 *   不如把页面图丢给多模态模型，让它把图上的字念出来 —— 一页一两秒，
 *   不需要下载任何模型，配一个地址就能用（也可以完全本地，见下）。
 *
 * 零依赖怎么拿到页面图：
 *   不渲染 PDF（那要 pdfium 这类原生库）。扫描件的每一页在 PDF 里本来就是
 *   一个**嵌进去的整页位图**，我们直接从 PDF 对象结构里把它抠出来：
 *     · /DCTDecode  —— 流里就是一张现成的 JPEG，原样拿来用
 *     · /FlateDecode —— 解压得到裸像素，用 Node 自带的 zlib 编成 PNG
 *   这条路对「扫描件」（也正是唯一需要它的场景）几乎总是有效。
 *
 * 和 MinerU 的分工：
 *   · 视觉模型：配好就能用，零模型下载，快。可以完全本地（Ollama）。
 *   · MinerU  ：全本地离线，但要装 800 MB 模型，慢。
 *   默认顺序是「先视觉、后 MinerU」。
 *
 * ⚠️ 隐私边界：走远程 API 时**页面图片会被发到模型厂商**。
 *    想完全不外传就用本地模型：GALGAME_VISION_BASE=http://127.0.0.1:11434/v1
 */

import zlib from 'node:zlib'

import { findObjects, dictOf, streamOf, matchAll } from './pdfbytes.js'

export const DEFAULT_MAX_PAGES = Number(process.env.GALGAME_VISION_MAX_PAGES || 40)
export const PAGE_TIMEOUT = Number(process.env.GALGAME_VISION_TIMEOUT || 180) * 1000

/** 小于这个尺寸的图是装饰/logo，不是正文页面 */
const MIN_PAGE_IMAGE_PIXELS = 250000

/** 让模型只吐正文，别加解释 —— 抽取不是聊天 */
export const DEFAULT_PROMPT = [
  '把这张书页图片里的文字**完整、逐字**转写成纯文本，保持原有段落顺序。',
  '要求：',
  '· 只输出正文，不要任何解释、不要加「以下是」「这张图片」之类的话；',
  '· 不要翻译、不要改写、不要总结；',
  '· 看不清的字用 □ 代替，不要猜；',
  '· 如果整页没有文字（比如纯插图页），只回一个空行。',
].join('\n')

/**
 * 读开关。**每次现读**，不要缓存成模块常量 ——
 * 缓存的话测试改不动它，用户在跑起来之后再设环境变量也不生效。
 */
export function visionMode() {
  return (process.env.GALGAME_VISION || 'auto').trim().toLowerCase()
}

export function visionConfig() {
  return {
    base: (process.env.GALGAME_VISION_BASE || '').trim().replace(/\/+$/, ''),
    model: (process.env.GALGAME_VISION_MODEL || '').trim(),
    key: (process.env.GALGAME_VISION_KEY || '').trim(),
    maxPages: DEFAULT_MAX_PAGES,
    timeout: PAGE_TIMEOUT,
    prompt: process.env.GALGAME_VISION_PROMPT || DEFAULT_PROMPT,
  }
}

export function visionConfigured() {
  const c = visionConfig()
  return Boolean(c.base && c.model)
}

/** 这一路能不能用：开关允许 + 地址模型都配了（图像抠取是纯 JS，永远可用）。 */
export function visionAvailable() {
  if (['0', 'off', 'false', 'no', 'none'].includes(visionMode())) return false
  return visionConfigured()
}

export function visionStatus() {
  const c = visionConfig()
  return {
    configured: visionConfigured(),
    renderer: true,               // 纯 JS，不依赖任何外部渲染器
    ready: visionAvailable(),
    base: c.base || null,
    model: c.model || null,
    extractor: 'embedded-image',
    maxPages: c.maxPages,
  }
}

/* ══════════════════════════════════════════════════════════════════
   PDF → 页面图片（纯 JS，零依赖）
   ══════════════════════════════════════════════════════════════════ */

const CRC_TABLE = (() => {
  const t = new Int32Array(256)
  for (let n = 0; n < 256; n++) {
    let c = n
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1
    t[n] = c
  }
  return t
})()

function crc32 (buf) {
  let c = -1
  for (let i = 0; i < buf.length; i++) c = CRC_TABLE[(c ^ buf[i]) & 0xff] ^ (c >>> 8)
  return (c ^ -1) >>> 0
}

function pngChunk (type, data) {
  const len = Buffer.alloc(4)
  len.writeUInt32BE(data.length)
  const body = Buffer.concat([Buffer.from(type, 'latin1'), data])
  const crc = Buffer.alloc(4)
  crc.writeUInt32BE(crc32(body))
  return Buffer.concat([len, body, crc])
}

/** 裸像素 → PNG（Node 自带 zlib 就够，不用任何图像库） */
export function encodePng (width, height, channels, raw) {
  const colorType = channels === 1 ? 0 : 2
  const stride = width * channels
  const rows = Buffer.alloc((stride + 1) * height)
  for (let y = 0; y < height; y++) {
    rows[y * (stride + 1)] = 0                     // filter: none
    raw.copy(rows, y * (stride + 1) + 1, y * stride, (y + 1) * stride)
  }
  const ihdr = Buffer.alloc(13)
  ihdr.writeUInt32BE(width, 0)
  ihdr.writeUInt32BE(height, 4)
  ihdr[8] = 8                                      // bit depth
  ihdr[9] = colorType
  return Buffer.concat([
    Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
    pngChunk('IHDR', ihdr),
    pngChunk('IDAT', zlib.deflateSync(rows, { level: 6 })),
    pngChunk('IEND', Buffer.alloc(0)),
  ])
}

/** 从 PDF 对象里抠页面图；返回 [{w,h,mime,data}]，按页序。 */
export function pageImages (buf, maxPages = DEFAULT_MAX_PAGES) {
  const s = buf.toString('latin1')
  const objs = findObjects(s)
  if (!objs.size) return { pages: [], note: '这个 PDF 没有可解析的对象结构' }

  const pages = []
  for (const [num, body] of objs) {
    const d = dictOf(body)
    if (/\/Type\s*\/Page\b/.test(d) && !/\/Type\s*\/Pages\b/.test(d)) pages.push([num, d])
  }
  pages.sort((a, b) => a[0] - b[0])
  if (!pages.length) return { pages: [], note: '没找到页面对象' }

  const out = []
  let unsupported = 0
  let tooSmall = 0

  for (const [, d] of pages) {
    if (out.length >= maxPages) break
    for (const num of xobjectNums(objs, d)) {
      const body = objs.get(num) || ''
      // streamOf 给的是 latin1 字符串（PDF 对象是当文本扫的），
      // 但图像流里全是二进制字节 —— 必须先按 latin1 还原成 Buffer，
      // 否则 base64 编码和 zlib 解压都会把 >127 的字节搞坏。
      const img = imageFromObject(dictOf(body), Buffer.from(streamOf(body), 'latin1'))
      if (!img) continue
      if (img.unsupported) { unsupported++; continue }
      if (img.w * img.h < MIN_PAGE_IMAGE_PIXELS) { tooSmall++; continue }
      out.push(img)
    }
  }

  let note = ''
  if (!out.length && unsupported) {
    note = '这本 PDF 的页面图是 ' + unsupported + ' 张 CCITT/JPEG2000 编码的（浏览器和模型都不认这两种），读不了'
  } else if (!out.length && tooSmall) {
    note = '只找到 ' + tooSmall + ' 张小图（logo/装饰），没有整页的正文图'
  }
  return { pages: out, note }
}

function xobjectNums (objs, pageDict) {
  let res = ''
  const rm = pageDict.match(/\/Resources\s*(\d+)\s+\d+\s+R/)
  if (rm) res = dictOf(objs.get(parseInt(rm[1], 10)) || '')
  else {
    const rm2 = pageDict.match(/\/Resources\s*<</)
    if (rm2) res = dictOf(pageDict.slice(rm2.index))
  }
  if (!res) return []
  let xo = ''
  const xm = res.match(/\/XObject\s*<<([\s\S]*?)>>/)
  if (xm) xo = xm[1]
  else {
    const xr = res.match(/\/XObject\s+(\d+)\s+\d+\s+R/)
    if (xr) xo = dictOf(objs.get(parseInt(xr[1], 10)) || '')
  }
  const nums = []
  for (const [, , num] of matchAll(xo, /\/([A-Za-z0-9#_.\-]+)\s+(\d+)\s+\d+\s+R/g)) {
    nums.push(parseInt(num, 10))
  }
  return nums
}

/** 一个图像 XObject → {w,h,mime,data}；认不出来返回 null 或 {unsupported:true} */
function imageFromObject (dict, rawStream) {
  if (!/\/Subtype\s*\/Image/.test(dict)) return null
  const w = parseInt((dict.match(/\/Width\s+(\d+)/) || [])[1] || '0', 10)
  const h = parseInt((dict.match(/\/Height\s+(\d+)/) || [])[1] || '0', 10)
  if (!w || !h) return null

  const filters = ((dict.match(/\/Filter\s*(\[[^\]]*\]|\/\w+)/) || [])[1] || '')
  if (/DCTDecode/.test(filters)) {
    // 流里就是一张现成的 JPEG —— 一个字节都不用解
    return { w, h, mime: 'image/jpeg', data: rawStream }
  }
  if (/CCITTFaxDecode|JPXDecode|JBIG2Decode/.test(filters)) {
    return { unsupported: true, w, h }
  }
  if (/FlateDecode/.test(filters)) {
    let raw
    try { raw = zlib.inflateSync(rawStream) } catch { try { raw = zlib.inflateRawSync(rawStream) } catch { return null } }
    const bpc = parseInt((dict.match(/\/BitsPerComponent\s+(\d+)/) || [])[1] || '8', 10)
    if (bpc !== 8) return { unsupported: true, w, h }          // 1/2/4 位的要靠解码器，跳过

    // 颜色空间要**精确**匹配：曾经写成 /DeviceGray|G/，而 "/DeviceRGB" 里也有个 G，
    // 结果把 RGB 当灰度编，图只用了 1/3 的像素、模型读出来是花的。
    const cs = ((dict.match(/\/ColorSpace\s*(\/\w+)/) || [])[1] || '/DeviceRGB')
    const channels = /^\/(DeviceGray|G)$/.test(cs) ? 1 : /^\/(DeviceCMYK|CMYK)$/.test(cs) ? 4 : 3
    if (channels === 4) return { unsupported: true, w, h }      // CMYK 要先转色，跳过

    raw = applyPredictor(raw, dict, w, channels)
    if (raw.length < w * h * channels) return null
    const pixels = raw.subarray(0, w * h * channels)
    return { w, h, mime: 'image/png', data: encodePng(w, h, channels, pixels) }
  }
  return null
}

/**
 * 还原 PNG/TIFF 预测器（PNG predictors）。
 *
 * 很多 PDF 里的 Flate 图像不是裸像素，而是「每行前面加一个 filter type 字节」的
 * PNG 预测编码。不还原的话拿到的就是一堆差分值 —— 编出来的 PNG 能打开，
 * 但内容全是噪声，模型只会读出一堆乱码。这个坑很隐蔽，必须处理。
 */
export function applyPredictor (raw, dict, width, channels) {
  const dp = dict.match(/\/DecodeParms\s*<<([\s\S]*?)>>/)
  if (!dp) return raw
  const pred = parseInt((dp[1].match(/\/Predictor\s+(\d+)/) || [])[1] || '1', 10)
  if (!(pred >= 10)) return raw                    // 1 = 没用预测器
  const colors = parseInt((dp[1].match(/\/Colors\s+(\d+)/) || [])[1] || String(channels), 10)
  const bpc = parseInt((dp[1].match(/\/BitsPerComponent\s+(\d+)/) || [])[1] || '8', 10)
  const columns = parseInt((dp[1].match(/\/Columns\s+(\d+)/) || [])[1] || String(width), 10)

  const bpp = Math.max(1, Math.ceil(colors * bpc / 8))      // 每像素字节数（预测的最小单位）
  const rowLen = Math.ceil(colors * bpc * columns / 8)
  const rows = Math.floor(raw.length / (rowLen + 1))
  if (!rows || !rowLen) return raw

  const out = Buffer.alloc(rowLen * rows)
  let prev = Buffer.alloc(rowLen)
  for (let r = 0; r < rows; r++) {
    const ft = raw[r * (rowLen + 1)]
    const row = Buffer.from(raw.subarray(r * (rowLen + 1) + 1, (r + 1) * (rowLen + 1)))
    if (ft === 1 || ft === 2 || ft === 3 || ft === 4) {
      for (let j = 0; j < rowLen; j++) {
        const a = j >= bpp ? row[j - bpp] : 0
        const b = prev[j]
        const c = j >= bpp ? prev[j - bpp] : 0
        let add = 0
        if (ft === 1) add = a
        else if (ft === 2) add = b
        else if (ft === 3) add = (a + b) >> 1
        else {
          const p = a + b - c
          const pa = Math.abs(p - a); const pb = Math.abs(p - b); const pc = Math.abs(p - c)
          add = (pa <= pb && pa <= pc) ? a : (pb <= pc ? b : c)
        }
        row[j] = (row[j] + add) & 0xff
      }
    }
    row.copy(out, r * rowLen)
    prev = row
  }
  return out
}

/* ══════════════════════════════════════════════════════════════════
   调用模型
   ══════════════════════════════════════════════════════════════════ */

/** 把一页图的文字念出来。 */
export async function transcribePage (img, cfg = visionConfig(), timeoutMs) {
  const b64 = img.data.toString('base64')
  const body = {
    model: cfg.model,
    temperature: 0,
    max_tokens: 8000,
    messages: [{
      role: 'user',
      content: [
        { type: 'text', text: cfg.prompt },
        { type: 'image_url', image_url: { url: `data:${img.mime};base64,${b64}` } },
      ],
    }],
  }
  const headers = { 'Content-Type': 'application/json' }
  if (cfg.key) headers.Authorization = 'Bearer ' + cfg.key

  const ac = new AbortController()
  const timer = setTimeout(() => ac.abort(), timeoutMs || cfg.timeout)
  try {
    const res = await fetch(cfg.base + '/chat/completions', {
      method: 'POST', headers, body: JSON.stringify(body), signal: ac.signal,
    })
    const text = await res.text()
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${text.slice(0, 200)}`)
    let j
    try { j = JSON.parse(text) } catch { throw new Error('返回的不是 JSON：' + text.slice(0, 120)) }
    const msg = j.choices && j.choices[0] && j.choices[0].message
    return ((msg && msg.content) || '').trim()
  } finally {
    clearTimeout(timer)
  }
}

/**
 * 读整本 PDF。返回 {text, pages, note}。
 * onProgress(i, n, info) 用来给界面报进度。
 */
export async function readPdf (buf, filename = '', onProgress) {
  const cfg = visionConfig()
  if (!visionAvailable()) return { text: '', pages: 0, note: '没配视觉模型' }

  const { pages, note } = pageImages(buf, cfg.maxPages)
  if (!pages.length) return { text: '', pages: 0, note: note || '这本书里没有可读的页面图' }

  const chunks = []
  let failed = 0
  for (let i = 0; i < pages.length; i++) {
    const img = pages[i]
    if (onProgress) onProgress(i + 1, pages.length, `${img.w}×${img.h}`)
    try {
      const t = await transcribePage(img, cfg)
      if (t) chunks.push(t)
    } catch (e) {
      failed++
      if (onProgress) onProgress(i + 1, pages.length, '失败：' + e.message.slice(0, 60))
      if (failed >= 3 && !chunks.length) {
        return { text: '', pages: pages.length, note: '连续几页都读失败：' + e.message.slice(0, 120) }
      }
    }
  }

  const text = chunks.join('\n\n')
  let out = ''
  if (failed) out = `${failed}/${pages.length} 页读取失败`
  return { text, pages: pages.length, note: out }
}

/* ══════════════════════════════════════════════════════════════════
   下面是给 textract.js 复用的 PDF 底层件（同一份实现，别各写一套）
   ══════════════════════════════════════════════════════════════════ */
